concordium-consensus/src/Concordium/GRPC2.hs

```diff
@@ -1574,6 +1575,59 @@ getInstanceInfoV2 cptr blockType blockHashPtr addrIndex addrSubindex outHash out
     res <- runMVR (Q.getInstanceInfo bhi caddr) mvr
     returnMessageWithBlock (copier outVec) outHash res
 
+-- |An opaque representation of the place where we write the mutable state in the 'getInstanceStateV2' query.
+data PersistentStateReceiver
+
+-- |A type of callback that copies the V1 contract persistent state into the
+-- provided recevier.
+type PersistentStateCopier = Ptr PersistentStateReceiver -> Ptr StateV1.PersistentState -> StateV1.LoadCallback -> IO ()
+
+-- |Boilerplate wrapper to invoke C callbacks.
+foreign import ccall "dynamic" callPersistentStateCopier :: FunPtr PersistentStateCopier -> PersistentStateCopier
+
+getInstanceStateV2 ::
+    StablePtr Ext.ConsensusRunner ->
+    -- |Block type.
+    Word8 ->
+    -- |Block hash.
+    Ptr Word8 ->
+    -- |Smart contract instance index.
+    Word64 ->
+    -- |Smart contract instance subindex.
+    Word64 ->
+    -- |Out pointer for writing the block hash that was used.
+    Ptr Word8 ->
+    -- |If the instance is a V0 instance its instance state is written to this vector.
+    -- Otherwise this is unused.
+    Ptr ReceiverVec ->
+    -- |Callback to output data in case it is a V0 instance.
+    FunPtr CopyToVecCallback ->
+    -- |If the instance is a V1 instance its instance state and callbacks are written to this vector.
```

Comment:
I suppose we could do that. But then we can't use the CopyToVecCallback since the types don't match.

We could just have the one callback with an additional argument and use the null pointers to signal which version we are writing. but I think that is more complicated than this.